### PR TITLE
feat(sentry): production monitoring for web and api

### PR DESCRIPTION
**Pull request by lehuygiang28**
## Summary
- Add Sentry foundation for **vkara-web** (Next.js) and **vkara-api** (Elysia): errors, traces, logs, replay, env isolation (`development` | `production`)
- Harden frontend: scrubbing, device/room context, middleware wrap, silent auto-recovery (soft reset → hard home)
- Ship ops dashboards + scripts; verify FE↔BE distributed tracing and recovery with unit/Playwright checks

## Test plan
- [x] Local: API `/health` + `/debug-sentry` (dev only) appear in Sentry `vkara-api`
- [x] Local: web client error + Replay in `vkara-web`; shared `trace_id` across FE/BE
- [x] Soft recovery: `/e2e-recovery?mode=soft` recovers without user click
- [x] Hard recovery: `/e2e-recovery?mode=hard` redirects to `/`
- [x] Dashboards load for development/production (Golden Signals, API Deep Dive, Web Client)
- [x] After merge: set Vercel Production env (`NEXT_PUBLIC_SENTRY_DSN`, `SENTRY_DSN`, `SENTRY_AUTH_TOKEN`, etc.) and confirm events on `vkara.vercel.app`

Made with [Cursor](https://cursor.com)

## Summary by Sourcery

Integrate Sentry-based monitoring and error handling across the web and API apps, including environment-aware configuration, logging, and health/recovery flows.

New Features:
- Add Sentry instrumentation for the API server, queues, and Redis clients, including cron monitoring, metrics, and centralized unexpected error capture.
- Introduce Sentry integration for the Next.js web app with client/server/edge configs, tracing, replay, and middleware wrapping, plus device/room context tagging.
- Expose a Sentry-backed `/health` endpoint and a gated `/debug-sentry` endpoint in the API for operational checks and error verification.
- Provide scripts and configurations to verify Sentry connectivity, create environment-scoped Sentry dashboards, and tune runtime behavior via env vars.

Enhancements:
- Wire Winston logging into Sentry with safe attribute flattening and redaction, and update log metadata for the API service.
- Improve robustness of background jobs and upstream integrations (YouTube, TikTok, room cleanup, hourly reports) with monitored Redis connections and Sentry-backed failure reporting.
- Refine error boundaries in the web app to auto-report crashes to Sentry and implement soft reset then hard navigation recovery flows, with E2E and unit tests validating behavior.
- Extend shared env packages with Sentry-specific configuration helpers and utilities for resolving environments, sample rates, log levels, and cron monitor allowlists.

Build:
- Wrap the Next.js config with `withSentryConfig`, configure source map upload and cleanup, and expose Vercel env metadata for Sentry tagging.

Deployment:
- Add shared Sentry environment and DSN configuration for API and web, suitable for Vercel-based production and development deployments.

Tests:
- Add unit tests for Sentry env utilities and error-boundary recovery planning, plus E2E-style scripts to verify Sentry wiring and auto-recovery flows.

### CODE REVIEW

File: apps/api/src/utils/logger.ts [hunk 1]
```diff
@@ -30,9 +33,55 @@ const customFormat = winston.format.printf(({ level, message, timestamp, context
...
+        }
+        const valueType = typeof value;
+        if (valueType === 'string' || valueType === 'number' || valueType === 'boolean') {
+            continue;
+        }
+        rest[key] = safeJson(value);
+    }
+
+    return rest as winston.Logform.TransformableInfo;
+})();
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/api/src/utils/logger.ts:76]: Fixed in d442a67 — mutate/return `info` in place so Winston Symbol keys are preserved.

File: apps/web/src/hooks/use-error-boundary-recovery.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        if (!raw) {
+            return { fingerprint, attempts: 0, lastAt: 0 };
+        }
+        const parsed = JSON.parse(raw) as RecoveryBucket;
+        if (
+            parsed.fingerprint !== fingerprint ||
+            Date.now() - parsed.lastAt > 5 * 60 * 1000
+        ) {
+            return { fingerprint, attempts: 0, lastAt: 0 };
+        }
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/src/hooks/use-error-boundary-recovery.ts]: Fixed in d442a67 — defensive object-shape check before reading bucket fields.

File: apps/api/src/sentry/health.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+export async function getServiceHealth(input: {
+    redis: Redis;
+    wsConnections: number;
+}): Promise<ServiceHealth> {
+    let redisStatus: 'up' | 'down' = 'down';
+    try {
+        redisStatus = (await input.redis.ping()) === 'PONG' ? 'up' : 'down';
+    } catch {
+        redisStatus = 'down';
+    }
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/api/src/sentry/health.ts]: Fixed in d442a67 — `Promise.race` with 2s timeout; hung Redis reports `down`.

File: apps/api/src/sentry/jobs.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    client.on('error', (error: Error) => {
+        redisLogger.error('Redis client error', { error, redis_client: name });
+        captureUnexpected(error, {
+            tags: { area: 'redis', redis_client: name },
+        });
+    });
+
+    return client;
+}
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/api/src/sentry/jobs.ts:40]: Fixed in d442a67 — 60s per-client cooldown on Sentry capture (logs still emit every error).

File: apps/web/src/lib/sentry/scrub.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * Used by client / server / edge inits (`sendDefaultPii` stays on for IP + UA).
+ */
+export function scrubSentryEvent(event: ErrorEvent, _hint?: EventHint): ErrorEvent | null {
+    if (event.request?.headers) {
+        redactRecord(event.request.headers as Record<string, unknown>);
+    }
+    if (event.request?.cookies) {
+        event.request.cookies = { '[Redacted]': '[Redacted]' };
+    }
+    if (event.request?.data && typeof event.request.data === 'object') {
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/src/lib/sentry/scrub.ts:27]: Fixed in d442a67 — keep cookie names, redact values only.

File: apps/web/src/hooks/use-error-boundary-recovery.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
+'use client';
+
+import { useEffect, useRef, useState } from 'react';
+import * as Sentry from '@sentry/nextjs';
+
+export const ERROR_RECOVERY_STORAGE_KEY = 'vkara_error_recovery';
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/src/hooks/use-error-boundary-recovery.ts:6]: Skipping — `planRecoveryAttempt` / bucket helpers are intentionally exported for unit tests and the e2e recovery harness.

File: apps/web/src/components/sentry/app-error-boundary.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    planRecoveryAttempt,
+} from '@/hooks/use-error-boundary-recovery';
+import { useI18n } from '@/locales/client';
+
+type FallbackProps = {
+    error: unknown;
+    resetError: () => void;
+};
+
+function AutoRecoverFallback({ error, resetError }: FallbackProps) {
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/src/components/sentry/app-error-boundary.tsx:22]: Fixed in d442a67 — `AutoRecoverFallback` now delegates to `useErrorBoundaryRecovery` (`boundaryTag: react_tree`).

File: scripts/sentry/create-dashboards.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    existingId?: string,
+) {
+    const body = {
+        title: dashboard.title,
+        widgets: dashboard.widgets,
+        projects: dashboard.projects ?? [-1],
+        period: dashboard.period ?? '24h',
+        environment: dashboard.environment,
+        isFavorited: dashboard.isFavorited ?? false,
+    };
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply scripts/sentry/create-dashboards.ts:86]: Fixed in d442a67 — `description` is included in the create/update request body.

File: apps/web/src/instrumentation-client.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    SENTRY_IGNORE_ERRORS,
+    scrubSentryEvent,
+} from '@/lib/sentry/scrub';
+
+const dsn = process.env.NEXT_PUBLIC_SENTRY_DSN;
+const sentryEnvironment = readSentryEnvironmentFromProcess({
+    preferPublic: true,
+    runtimeHost: typeof window !== 'undefined' ? window.location.hostname : undefined,
+});
+const enabled = isSentryEnabled(dsn, process.env.SENTRY_ENABLED);
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/src/instrumentation-client.ts]: Fixed in d442a67 — client uses `NEXT_PUBLIC_SENTRY_ENABLED ?? SENTRY_ENABLED`; wired through `web-public` + `env.ts`.

File: apps/api/src/utils/logger.ts [hunk 10]
```diff
@@ -30,9 +33,55 @@ const customFormat = winston.format.printf(({ level, message, timestamp, context
...
+            continue;
+        }
+        const valueType = typeof value;
+        if (valueType === 'string' || valueType === 'number' || valueType === 'boolean') {
+            continue;
+        }
+        rest[key] = safeJson(value);
+    }
+
+    return rest as winston.Logform.TransformableInfo;
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/api/src/utils/logger.ts]: Fixed in d442a67 — same as Gemini finding; format mutates and returns `info`.

File: apps/api/src/server.ts [hunk 11]
```diff
@@ -38,40 +39,60 @@ export function sendToClient(ws: ElysiaWS, message: ServerMessage): void {
...
+            });
     })
     .on('stop', async () => {
         serverLogger.info('Server stop initiated');
         try {
             await shutdownTikTokPool().catch(() => {});
             await redis.quit();
+            await Sentry.close(2000);
             await wsServer.stop();
             serverLogger.info('Server stopped successfully');
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/api/src/server.ts:91]: Fixed in d442a67 — signal handler only calls `wsServer.stop()`; cleanup lives solely in the `stop` lifecycle (no nested `stop()`).

File: scripts/sentry/test-error-recovery.mjs [hunk 12]
```diff
@@ -0,0 +1,113 @@
+import { chromium } from '../../node_modules/.bun/playwright@1.60.0/node_modules/playwright/index.js';
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply scripts/sentry/test-error-recovery.mjs]: Fixed in d442a67 — resolve `playwright` via `createRequire` from `apps/api` (no Bun store version pin).

File: apps/web/sentry.server.config.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    dsn,
+    enabled,
+    environment: sentryEnvironment,
+    ...(release ? { release } : {}),
+    sendDefaultPii: true,
+    tracesSampleRate: resolveSentryTracesSampleRate(
+        process.env.SENTRY_TRACES_SAMPLE_RATE,
+        sentryEnvironment,
+    ),
+    includeLocalVariables: true,
```
[the remark this answers is not in the document]
Comment (lehuygiang28) [reply apps/web/sentry.server.config.ts]: Fixed in d442a67 — `includeLocalVariables` gated to `development` only.
